src/ur5e_arm.cpp

```diff
@@ -403,6 +403,54 @@ void UR5eArm::send_trajectory(const std::vector<vector6d_t>& p_p,
 
 // helper function to read a data packet and send a noop message
 void UR5eArm::read_and_noop() {
+    // check to see if an estop has occurred.
+    std::string status;
+    dashboard->commandSafetyStatus(status);
+
+    if (status.find(urcl::safetyStatusString(urcl::SafetyStatus::NORMAL)) == std::string::npos) {
+        // the arm is currently estopped. save this state.
+        estop.store(true);
+    } else {
+        // the arm is in a normal state.
+        if (estop.load()) {
+            // if the arm was previously estopped, attempt to recover from the estop.
+            // We should not enter this code without the user interacting with the arm in some way(i.e. resetting the estop)
+            try {
+                BOOST_LOG_TRIVIAL(info) << "recovering from e-stop" << std::endl;
+                driver->resetRTDEClient(path_offset + OUTPUT_RECIPE, path_offset + INPUT_RECIPE);
+
+                if (!dashboard->commandPowerOff()) {
+                    throw std::runtime_error("couldn't power off arm");
+                }
+                if (!dashboard->commandPowerOn()) {
+                    throw std::runtime_error("couldn't power on arm");
+                }
+                // Release the brakes
+                if (!dashboard->commandBrakeRelease()) {
+                    throw std::runtime_error("couldn't release the arm brakes");
+                }
+                BOOST_LOG_TRIVIAL(info) << "arm restarted, sending control program again" << std::endl;
+
+                // send control script
+                if (!driver->sendRobotProgram()) {
+                    throw std::runtime_error("failed to resend control program");
+                }
+                BOOST_LOG_TRIVIAL(info) << "send robot program successful, restarting communication" << std::endl;
+
+                // clear any currently running trajectory.
+                trajectory_running = false;
```

Comment:
we might want to do this when the estop is first detected. unsure of ideal behavior.

Reply:
I think that makes sense?  Likely the trajectory has already been cancelled, but best for us to track that state internally as accurately as possible.

Reply:
nope. when the estop is triggered we receive nothing from reportTrajectoryState, so I have to manage this myself.